Add OTA reflashing support on TX2

atharvanan1 opened the pull request:
Hey,

I tried to integrate the OTA setup highlighted here - https://github.com/OE4T/meta-tegra/wiki/Over-the-air-reflashing-process

**Setup Information**:
Pre-update Software built from - https://github.com/madisongh/test-distro/tree/warrior
`sysinstall-upgrader-initramfs` with modified `tegra-sysinstall-tools` as per https://github.com/BoulderAI/tegra-sysinstall/tree/add-UDA-support
Update Software built from - this PR

- Added `meta-tegra-support/classes/image_types_tegra.bbclass` to replace the `meta-tegra/classes/image_types_tegra.bbclass` -> this is done to change the PPTSIZE to 16896 as [required by the OTA process](https://github.com/OE4T/meta-tegra/wiki/Over-the-air-reflashing-process#key-considerations).
- Added `recipes-devtools` folder from https://github.com/madisongh/tegra-test-distro/tree/dunfell-l4t-r32.5.0/layers/meta-testdistro/recipes-devtools
- Added `packgegroup-demo-base.bbappend` from https://github.com/madisongh/tegra-test-distro/blob/dunfell-l4t-r32.5.0/layers/meta-testdistro/recipes-testdistro/packagegroups/packagegroup-demo-base.bbappend

---

`sysinstall-partition-layout` pipes out `partition_table` file as

```
1,APP,25376768,,,,
2,mts-bootpack,8192,,,,
3,mts-bootpack_b,8192,,,,
4,cpu-bootloader,1024,,,,
5,cpu-bootloader_b,1024,,,,
6,bootloader-dtb,1024,,,,
7,bootloader-dtb_b,1024,,,,
8,secure-os,6144,,,,
9,secure-os_b,6144,,,,
10,eks,4096,,,,
11,adsp-fw,8192,,,,
12,adsp-fw_b,8192,,,,
13,bpmp-fw,1208,,,,
14,bpmp-fw_b,1208,,,,
15,bpmp-fw-dtb,2048,,,,
16,bpmp-fw-dtb_b,2048,,,,
17,sce-fw,4096,,,,
18,sce-fw_b,4096,,,,
19,sc7,12288,,,,
20,sc7_b,12288,,,,
21,fusebypass,4096,,,,
22,BMP,262144,,,,
23,BMP_b,262144,,,,
24,recovery,129024,,,,
25,recovery-dtb,1024,,,,
26,kernel-bootctrl,512,,,,
27,kernel-bootctrl_b,512,,,,
28,kernel,163840,,,,
29,kernel_b,163840,,,,
30,kernel-dtb,1024,,,,
31,kernel-dtb_b,1024,,,,
32,RECROOTFS,614400,,,,
33,APP_b,25376768,,,,
34,UDA,REMAIN,,,,
```

---

Below is the log from after partitioning gpt disk

```Disk /dev/mmcblk0: 61071360 sectors, 29.1 GiB
Sector size (logical/physical): 512/512 bytes
Disk identifier (GUID): BE2DDE25-23BF-49A3-BA2F-6807D1016D24
Partition table holds up to 128 entries
Main partition table begins at sector 2 and ends at sector 33
First usable sector is 34, last usable sector is 61071326
Partitions will be aligned on 2-sector boundaries
Total free space is 0 sectors (0 bytes)

Number  Start (sector)    End (sector)  Size       Code  Name
1              34        25376801   12.1 GiB    0700  APP
2        25376802        25384993   4.0 MiB     0700  mts-bootpack
3        25384994        25393185   4.0 MiB     0700  mts-bootpack_b
4        25393186        25394209   512.0 KiB   0700  cpu-bootloader
5        25394210        25395233   512.0 KiB   0700  cpu-bootloader_b
6        25395234        25396257   512.0 KiB   0700  bootloader-dtb
7        25396258        25397281   512.0 KiB   0700  bootloader-dtb_b
8        25397282        25403425   3.0 MiB     0700  secure-os
9        25403426        25409569   3.0 MiB     0700  secure-os_b
10        25409570        25413665   2.0 MiB     0700  eks
11        25413666        25421857   4.0 MiB     0700  adsp-fw
12        25421858        25430049   4.0 MiB     0700  adsp-fw_b
13        25430050        25431257   604.0 KiB   0700  bpmp-fw
14        25431258        25432465   604.0 KiB   0700  bpmp-fw_b
15        25432466        25434513   1024.0 KiB  0700  bpmp-fw-dtb
16        25434514        25436561   1024.0 KiB  0700  bpmp-fw-dtb_b
17        25436562        25440657   2.0 MiB     0700  sce-fw
18        25440658        25444753   2.0 MiB     0700  sce-fw_b
19        25444754        25457041   6.0 MiB     0700  sc7
20        25457042        25469329   6.0 MiB     0700  sc7_b
21        25469330        25473425   2.0 MiB     0700  fusebypass
22        25473426        25735569   128.0 MiB   0700  BMP
23        25735570        25997713   128.0 MiB   0700  BMP_b
24        25997714        26126737   63.0 MiB    0700  recovery
25        26126738        26127761   512.0 KiB   0700  recovery-dtb
26        26127762        26128273   256.0 KiB   0700  kernel-bootctrl
27        26128274        26128785   256.0 KiB   0700  kernel-bootctrl_b
28        26128786        26292625   80.0 MiB    0700  kernel
29        26292626        26456465   80.0 MiB    0700  kernel_b
30        26456466        26457489   512.0 KiB   0700  kernel-dtb
31        26457490        26458513   512.0 KiB   0700  kernel-dtb_b
32        26458514        27072913   300.0 MiB   0700  RECROOTFS
33        27072914        52449681   12.1 GiB    0700  APP_b
34        52449682        61071326   4.1 GiB     0700  UDA
[  244.096703]  mmcblk0: p1 p2 p3 p4 p5 p6 p7 p8 p9 p10 p11 p12 p13 p14 p15 p16 p17 p18 p19 p20 p21 p22 p23 p24 p25 p26 p27 p28 p29 p30 p31 p32 p33 p34
[  244.172446]  mmcblk0: p1 p2 p3 p4 p5 p6 p7 p8 p9 p10 p11 p12 p13 p14 p15 p16 p17 p18 p19 p20 p21 p22 p23 p24 p25 p26 p27 p28 p29 p30 p31 p32 p33 p34
NOTE: secure boot not enabled, skipping partition encryption
UDA (/dev/mmcblk0p34): formatting (-i 4096)...[OK]
[  249.299200] EXT4-fs (mmcblk0p34): mounted filesystem with ordered data mode. Opts: (null)
Secure boot not enabled; skipping machine ID setting
Installing rootfs to /dev/mmcblk0p1...
[  253.464004] EXT4-fs (mmcblk0p1): mounted filesystem with ordered data mode. Opts: (null)
[  267.262992] EXT4-fs (mmcblk0p1): mounted filesystem with ordered data mode. Opts: (null)
APP: clean, 3835/3172288 files, 320570/3172096 blocks
Installing rootfs to /dev/mmcblk0p33...
[  271.268307] EXT4-fs (mmcblk0p33): mounted filesystem with ordered data mode. Opts: (null)
[  283.591877] EXT4-fs (mmcblk0p33): mounted filesystem with ordered data mode. Opts: (null)
APP_b: clean, 3835/3172288 files, 320569/3172096 blocks
[  283.718564] EXT4-fs (mmcblk0p1): mounted filesystem with ordered data mode. Opts: (null)
Updating bootloader...
Clearing boot device /dev/mmcblk0boot0... /dev/mmcblk0boot1... [OK]
Processing bpmp-fw... [OK]
Processing bpmp-fw_b... [OK]
Processing sce-fw... [OK]
Processing sce-fw_b... [OK]
Processing cpu-bootloader... [OK]
Processing cpu-bootloader_b... [OK]
Processing spe-fw... [OK]
Processing spe-fw_b... [OK]
Processing secure-os... [OK]
Processing secure-os_b... [OK]
Processing adsp-fw... [OK]
Processing adsp-fw_b... [OK]
Processing badpage-fw... [OK]
Processing badpage-fw_b... [OK]
Processing secure-os... [OK]
Processing secure-os_b... [OK]
Processing BMP... [OK]
Processing BMP_b... [OK]
Processing mts-bootpack... [OK]
Processing mts-bootpack_b... [OK]
Processing mts-preboot... [OK]
Processing mts-preboot_b... [OK]
Processing sc7... [OK]
Processing sc7_b... [OK]
Processing bpmp-fw-dtb... [OK]
Processing bpmp-fw-dtb_b... [OK]
Processing bootloader-dtb... [OK]
Processing bootloader-dtb_b... [OK]
Processing VER... [OK]
Processing VER_b... [OK]
Processing MB1_BCT... [OK]
Processing MB1_BCT_b... [OK]
Processing kernel... [OK]
Processing kernel_b... [OK]
Processing kernel-dtb... [OK]
Processing kernel-dtb_b... [OK]
Processing mb2... [OK]
Processing mb2_b... [OK]
Processing BCT... [offset=3584]...[offset=16384]...[offset=0]...[OK]
Processing mb1... [OK]
Processing mb1_b... [OK]
Processing eks... [OK]
Processing dram-ecc-fw... [OK]
Slot 0 marked as active for next boot
PASS: successful installation of demo-image-base_3.1-9eadaec-atharva
```

Upon reboot the bootloader fails with this:

```[0000.174] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.162] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0[0000.162] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.151] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.140] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.140] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.140] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.154] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.151] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.140] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.140] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0[0000.140] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.140] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.140] C> ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
[0�[0000.062] C> Failed to update boot chain
[0000.066] C> ERROR: Highest Layer Module = 0x54, Lowest Layer Module = 0x54,
Aux Info = 0x1, Reason = 0x2```

ichergui reviewed (commented):
Hey @atharvanan1 
Thanks for sharing this PR, I guess that you tested this only with `Jetson TX2`
Do you need support/help to validate this with other `Jetson's platform` ?

I'm seeing some errors in the last traces ? Do you need help to fix these errors ?

madisongh commented:
@atharvanan1 Looks like this is due to a difference in the NVIDIA flashing tools between L4T R32.4.3 and R32.5.x.  If I directly flash your R32.4.3 image onto my TX2, this is what I see for the eMMC partitioning:
```
Number  Start (sector)    End (sector)  Size       Code  Name
   1              40        25376807   12.1 GiB    0700  APP
   2        25376808        25384999   4.0 MiB     0700  mts-bootpack
   3        25385000        25393191   4.0 MiB     0700  mts-bootpack_b
   4        25393192        25394215   512.0 KiB   0700  cpu-bootloader
   5        25394216        25395239   512.0 KiB   0700  cpu-bootloader_b
   6        25395240        25396263   512.0 KiB   0700  bootloader-dtb
   7        25396264        25397287   512.0 KiB   0700  bootloader-dtb_b
   8        25397288        25403431   3.0 MiB     0700  secure-os
   9        25403432        25409575   3.0 MiB     0700  secure-os_b
  10        25409576        25413671   2.0 MiB     0700  eks
  11        25413672        25421863   4.0 MiB     0700  adsp-fw
  12        25421864        25430055   4.0 MiB     0700  adsp-fw_b
  13        25430056        25431263   604.0 KiB   0700  bpmp-fw
  14        25431264        25432471   604.0 KiB   0700  bpmp-fw_b
  15        25432472        25434519   1024.0 KiB  0700  bpmp-fw-dtb
  16        25434520        25436567   1024.0 KiB  0700  bpmp-fw-dtb_b
  17        25436568        25440663   2.0 MiB     0700  sce-fw
  18        25440664        25444759   2.0 MiB     0700  sce-fw_b
  19        25444760        25457047   6.0 MiB     0700  sc7
  20        25457048        25469335   6.0 MiB     0700  sc7_b
  21        25469336        25473431   2.0 MiB     0700  fusebypass
  22        25473432        25735575   128.0 MiB   0700  BMP
  23        25735576        25997719   128.0 MiB   0700  BMP_b
  24        25997720        26126743   63.0 MiB    0700  recovery
  25        26126744        26127767   512.0 KiB   0700  recovery-dtb
  26        26127768        26128279   256.0 KiB   0700  kernel-bootctrl
  27        26128280        26128791   256.0 KiB   0700  kernel-bootctrl_b
  28        26128792        26292631   80.0 MiB    0700  kernel
  29        26292632        26456471   80.0 MiB    0700  kernel_b
  30        26456472        26457495   512.0 KiB   0700  kernel-dtb
  31        26457496        26458519   512.0 KiB   0700  kernel-dtb_b
  32        26458520        27072919   300.0 MiB   0700  RECROOTFS
  33        27072920        52449687   12.1 GiB    0700  APP_b
  34        52449688        61071326   4.1 GiB     0700  UDA
```
Note that the starting sector for `APP` is `40` rather than `34`.  It looks like the NVIDIA flashing tools are forcing a sector alignment of 8 sectors (4KiB) for partitions in R32.4.3, which isn't there in R32.5.x.

I'll update `tegra-sysinstall-tools` to add the UDA partition and to set the partition alignment default to 8.

madisongh commented:
@atharvanan1 I've updated `tegra-sysinstall-tools`, and have updated my test distros with a recipe for the new version (1.6.0).

atharvanan1 commented:
@madisongh Great! Thanks for your input! I'll try it out on my end, and report back with my findings.

@ichergui Thanks for the offer. I created this PR so that it's easier for Matt to review my testing with OTA flashing update.

madisongh commented:
I was wrong about the root cause being a difference in the L4T tools....  The actual root cause is due to the `<align_boundary>` element for the APP partition in the XML file, with that partition appearing first in your flash layout but not in the ones I was using (where I reordered the physical layout to move the APP partition to after all the boot-related partitions).

But since APP-first is more typically found, I'll leave things as they are for now, then make sure that when I generate the `partition_table` file I correctly account for the `<align_boundary>` setting.

atharvanan1 commented:
> Hey @atharvanan1
> Thanks for sharing this PR, I guess that you tested this only with `Jetson TX2`
> Do you need support/help to validate this with other `Jetson's platform` ?
> 
> I'm seeing some errors in the last traces ? Do you need help to fix these errors ?

@ichergui Let's keep this PR for TX2 device. With Matt's setup, I'm testing on TX2 c-boot setup, and will be porting to TX2 uboot setup from L4T-28.2. I'll update the status of my testing and add some additional information on what I've tested so far on https://github.com/OE4T/meta-tegra/wiki/Over-the-air-reflashing-process

ichergui commented:
> > Hey @atharvanan1
> > Thanks for sharing this PR, I guess that you tested this only with `Jetson TX2`
> > Do you need support/help to validate this with other `Jetson's platform` ?
> > I'm seeing some errors in the last traces ? Do you need help to fix these errors ?
> 
> @ichergui Let's keep this PR for TX2 device. With Matt's setup, I'm testing on TX2 c-boot setup, and will be porting to TX2 uboot setup from L4T-28.2. I'll update the status of my testing and add some additional information on what I've tested so far on https://github.com/OE4T/meta-tegra/wiki/Over-the-air-reflashing-process

Perfect.

Thanks @atharvanan1 

Let me know if you need help

atharvanan1 commented:
@madisongh Thanks for that fix! The update process works with `tegra-demo-distro:dunfell-l4t-r32.4.3` branch. I'll update the commit to add any changes that are needed and document it well.

I was wondering if we could remove the dependency on PPTSIZE being 16896. If we could use partition_table file to get PPTSIZE that was used by the build and try to flash according to that. Are there any other dependencies we would want to watch out for?

madisongh commented:
> I was wondering if we could remove the dependency on PPTSIZE being 16896. If we could use partition_table file to get PPTSIZE that was used by the build and try to flash according to that. Are there any other dependencies we would want to watch out for?

That should be doable, as long as the actual size is 16896 or greater (if it's less, normal Linux GPT tools won't work).  There's already a field in the partition_table file for specifying the starting sector number, and that could be calculated from the info in the XML file.  I'll see if I can work that in along with handling the alignment tags.

atharvanan1 commented:
> as long as the actual size is 16896 or greater

Sounds good! I was thinking about PPTSIZE set up by [image_types_tegra.bbclass](https://github.com/OE4T/meta-tegra/blob/6772b7730a024868b7b3d613d3eebf51b30c0ef7/classes/image_types_tegra.bbclass#L234) when I said that. Let me know once the changes are done, I'll help test that on my end.

madisongh commented:
> Let me know once the changes are done, I'll help test that on my end.

OK, the update to `nvflashxmlparse` is in meta-tegra in all current branches, and I've updated tegra-sysinstall-tools to v1.6.1 to revert the default alignment change which is no longer required. You should be able to drop your `image_types_tegra.bbclass` replacement.

atharvanan1 commented:
**Tested OTA update:**
custom-sumo-l4t-r28.2 to dunfell-l4t-r32.4.3

**Issues**:
`/data` partition set up previous distro gets mounted while executing `tegra-sysinstall`. This creates a weird setup where you have a partition mounted, while we are trying to repartition the mmcblk0 device.

Apart from the above-mentioned issue, the update process worked fine. I had to interrupt the script, unmount /data, and restart the script.

Additional note: You can also use USB stick to deliver the files if the USB stick has ext4 filesystem.

madisongh commented on layers/meta-tegra-support/recipes-ota/packagegroups/packagegroup-demo-base.bbappend:
```diff
@@ -0,0 +1,8 @@
+    tegra-eeprom-tool \
+    tegra-fuse-tool \   
+    tegra-sysinstall-tools \
+"
```
I'm not sure we really want to include all of these packages in every image.  Can we drop this bbappend?

madisongh commented on layers/meta-tegra-support/recipes-ota/tegra-bup-payload/tegra-bup-payload_%.bbappend:
```diff
@@ -0,0 +1,3 @@
+RDEPENDS_${PN}_remove = "tegra-redundant-boot"
+RDEPENDS_${PN}_append = " tegra-boot-tools-updater"
+
```
Here, too, this will affect all images that include a BUP payload. I'm generally OK with this, I guess. I had held off on doing this in the dunfell-based branches, but did this in gatesgarth and later. You should review the changes I made in the gatesgarth branch over the Xmas holiday to see if they might also be needed:
```
$ git log --reverse --pretty=oneline --since=25-dec-2020 --before=4-jan-2021 origin/gatesgarth
```

atharvanan1 replied:
> I had held off on doing this in the dunfell-based branches
Was there a reason for doing this?

madisongh replied:
I just didn't want to run the risk of introducing new tools into the LTS branches.

atharvanan1 replied:
I agree with that. Is there any other solutions you might think of for this? 

I was wondering if we could include tegra-bootloader-update as part of sysinstall-upgrader-initramfs.

madisongh replied:
Well, the sysinstall scripts chroot into the installed rootfs to run the bootloader update, so that the tools/config/payload are all for the new OS rather than the old OS under which the upgrader initramfs is built... and I think it's probably safer to keep it that way.

I'm OK with making these changes now, since automating the OTA reflashing process is a useful new feature.

atharvanan1 replied:
Alright

> You should review the changes I made in the gatesgarth branch over the Xmas holiday to see if they might also be needed

I'll take a look at this and get back.

madisongh replied:
And FYI... @ichergui noticed that the changes I made earlier to the BUP payload contents cause failures when doing regular Mender-type updates (not reflashing) on the TX2.  The L4T `nv_update_engine` wasn't happy with the additions I made, so I've had to back them out and create a separate payload that can be used for a full initialization.

I've updated the test distros I mentioned on the wiki with the latest changes (including an update to the tegra-sysinstall scripts) to handle this.

atharvanan1 replied:
I have updated the branch to add whatever you have added on gatesgarth branch. I cross referenced the commits that are already added in `meta-mender-community`, updated with changes that were added for using `tegra-bootloader-update` tool.

madisongh replied:
There's one more patch for the update scripts [here](https://github.com/madisongh/tegra-test-distro/commit/fdd6ee50c27f849b19b9a9bd24b9667d1c9abac8#diff-01e509973933159d027521681546fd193965ce207cb7e20cadbd19400440492a) that needs to be ported over, but otherwise this looks good.

madisongh commented on layers/meta-tegra-support/recipes-ota/tegra-tools/tegra-sysinstall-tools_1.6.1.bb:
```diff
@@ -0,0 +1,24 @@
+
+FILES_${PN} += "${datadir}/tegra-sysinstall"
+EXTRA_RDEPENDS = ""
+EXTRA_RDEPENDS_cryptparts = "cryptsetup keystore-tools"
```
The above is specific to my test distro, so you should remove the EXTRA_RDEPENDS references here.

madisongh commented on layers/meta-tegra-support/recipes-ota/tegra-tools/tegra-sysinstall-tools_git.bb:
```diff
@@ -0,0 +1,27 @@
+                  bash curl util-linux-blkid util-linux-lsblk util-linux-mountpoint \
+                  parted gptfdisk e2fsprogs util-linux-mkfs util-linux-mount \
+                  util-linux-umount tegra-fuse-tool tegra-eeprom-tool-boardspec \
+                  sysinstall-partition-layout mtd-utils"
```
You can drop this recipe altogether.  The versioned one is the right one to include.

atharvanan1 commented:
I'll be running mender-torture-tests on `jetson-tx2`, `jetson-xavier-nx-emmc`, `jetson-nano-emmc` devices. I'll keep you posted with the results.

atharvanan1 commented:
Errors while running on `jetson-tx2`:
```
RuntimeError: machine ID changed
 during mender update
```
Errors while running on `jetson-tx2-uboot`:
```
RuntimeError: machine ID changed
 during mender update
```
---
Errors while running on `jetson-xavier-nx-devkit-emmc`:
```
RuntimeError: machine ID changed
 during mender update
```
---
Errors while running on `jetson-nano-emmc`:
```
time="2021-06-11T17:30:19Z" level=info msg="Collected output (stderr) while running script /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\nERR: cannot perform bootloader update\n\n---------- end of script output"
time="2021-06-11T17:30:19Z" level=error msg="ArtifactInstall_Leave script failed: statescript: error executing 'ArtifactInstall_Leave_80_bl-update': 1 : exit status 1"
Rolling back Artifact...
time="2021-06-11T17:30:19Z" level=info msg="Rolling back to the active partition: (1)."
time="2021-06-11T17:30:19Z" level=error msg="statescript: error executing 'ArtifactInstall_Leave_80_bl-update': 1 : exit status 1"
```

atharvanan1 commented:
From my testing,

1. Jetson TX2 (cboot) is successfully updated several times
2. Jetson Xavier NX is successfully updated several times

Apparently, jetson-nano-emmc device fails to update the bootloader. Probably, [this applies.](https://github.com/OE4T/tegra-demo-distro/blob/0e2b2d4324508972e0b7010f242dfd23a7528766/layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-install-script-uboot#L60)

ichergui commented:
> From my testing,
> 
> 1. Jetson TX2 (cboot) is successfully updated several times
> 2. Jetson Xavier NX is successfully updated several times
> 
> Apparently, jetson-nano-emmc device fails to update the bootloader. Probably, [this applies.](https://github.com/OE4T/tegra-demo-distro/blob/0e2b2d4324508972e0b7010f242dfd23a7528766/layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-install-script-uboot#L60)

Could you please share the logs ?

atharvanan1 commented:
Here's the log for 
TX2 mender-torture which is currently at 90th iteration or so - [jetson-tx2-mender-torture-90.log](https://github.com/OE4T/tegra-demo-distro/files/6645263/jetson-tx2-mender-torture-90.log)
Xavier-nx mender-torture where I tested with 20 mender updates and then a reboot torture - 
[xavier-mender-torture.log](https://github.com/OE4T/tegra-demo-distro/files/6645265/xavier-mender-torture.log)
[xavier-nx-reboot-torture.log](https://github.com/OE4T/tegra-demo-distro/files/6645266/xavier-nx-reboot-torture.log)

I know this isn't sufficient, however I feel that this is good enough to call successful. Let me know if you want me to test more. 

Errors with nano:
[jetson-nano-mender-torture.log](https://github.com/OE4T/tegra-demo-distro/files/6645268/jetson-nano-mender-torture.log)

atharvanan1 commented:
@ichergui @madisongh I have couple of questions and might need help on this.

It seems that how we handle VER partittions for `jetson-nano-emmc` has changed between [gatesgarth](https://github.com/OE4T/tegra-demo-distro/blob/gatesgarth/layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-bsp/tegra-binaries/mender-custom-flash-layout/jetson-nano-emmc/flash_mender.xml) (and what the tegra-bootloader-update expects) and [dunfell](https://github.com/mendersoftware/meta-mender-community/blob/dunfell/meta-mender-tegra/recipes-bsp/tegra-binaries/mender-custom-flash-layout/jetson-nano-emmc/flash_mender.xml). How do we address this? Should we not include the change from nv_update_engine to tegra-bootloader-update tool in `jetson-nano-emmc` device?

atharvanan1 commented:
Based on further investigations,

I found that the errors with `jetson-nano-emmc` are being caused by `tegra-bootloader-update` thinking something went wrong with VER and VER_b partitions - https://github.com/OE4T/tegra-boot-tools/blob/5a3e698f4f8158a14979df067f29cc26bbd47f75/tegra-bootloader-update.c#L930

Here's a hexdump for VER and VER_b partitions from `/dev/mmcblk0boot1` device on an image built off on `dunfell-l4t-r32.4.3` tegrademo-mender: (VER partitions are writted with zeros!)
```
03bb000 6b46 d969 6102 746c 6f62 746f 6d63 3d64
03bb010 7572 206e 656d 646e 7265 615f 746c 6f62
03bb020 746f 6d63 3b64 7220 6e75 6220 6f6f 6374
03bb030 646d 6100 6372 3d68 7261 006d 6162 6475
*
03bcb60 6564 5f76 6f66 5f72 6f62 746f 705f 7261
03bcb70 3b74 6620 0069 6576 646e 726f 6e3d 6976
03bcb80 6964 0061 0000 0000 0000 0000 0000 0000
03bcb90 0000 0000 0000 0000 0000 0000 0000 0000
*
03db000 9153 ce0c 6101 746c 6f62 746f 6d63 3d64
03db010 7572 206e 656d 646e 7265 615f 746c 6f62
03db020 746f 6d63 3b64 7220 6e75 6220 6f6f 6374
03db030 646d 6100 6372 3d68 7261 006d 6162 647
*
03dcb20 6373 6e61 645f 7665 665f 726f 625f 6f6f
03dcb30 5f74 6170 7472 203b 6966 7600 6e65 6f64
03dcb40 3d72 766e 6469 6169 0000 0000 0000 0000
03dcb50 0000 0000 0000 0000 0000 0000 0000 0000
*
0400000
```

Here's a hexdump from VER and VER_b partitions from `/dev/mmcblk0boot1` device on an image built off `dunfell-l4t-r32.4.3` branch: (VER partitions are intact)
```
*
03e0000 564e 0a33 2023 3352 2032 202c 4552 4956
03e0010 4953 4e4f 203a 2e34 0a33 4f42 5241 4944
03e0020 3d44 3433 3834 4220 414f 4452 4b53 3d55
03e0030 3030 3230 4620 4241 343d 3030 320a 3230
03e0040 3031 3136 3137 3030 3539 0a31 5942 4554
03e0050 3a53 3637 4320 4352 3233 323a 3430 3933
03e0060 3531 3536 0a35 0000 0000 0000 0000 0000
03e0070 0000 0000 0000 0000 0000 0000 0000 0000
*
03f0000 564e 0a33 2023 3352 2032 202c 4552 4956
03f0010 4953 4e4f 203a 2e34 0a33 4f42 5241 4944
03f0020 3d44 3433 3834 4220 414f 4452 4b53 3d55
03f0030 3030 3230 4620 4241 343d 3030 320a 3230
03f0040 3031 3136 3137 3030 3539 0a31 5942 4554
03f0050 3a53 3637 4320 4352 3233 323a 3430 3933
03f0060 3531 3536 0a35 0000 0000 0000 0000 0000
03f0070 0000 0000 0000 0000 0000 0000 0000 0000
*
0400000
```
---

Based on [Mender Tegra Partition Layout](https://docs.google.com/spreadsheets/d/14zVltS2LTwCNywoje6_8wzFzHQ3sYVhY6C7T0DkOnx0/edit#gid=932415434), we see that the UBENV is located around [0x3bb000 on /dev/mmcblk0boot1 device](https://docs.google.com/spreadsheets/d/1hHOA8RXyGx1o2L_syXeKm5VlxAZtGn38LwHEKQD2ySY/edit#gid=932415434&range=M58). 

fw_env.config
```
/dev/mmcblk0boot1 0x3bb000 0x20000
/dev/mmcblk0boot1 0x3db000 0x20000
```
0x3bb000 + 0x20000 = 0x3db000
0x3db000 + 0x20000 = 0x3fb000

Since, VER_b is located at 0x3e0000 and VER is located at 0x3f0000. They will get overridden. 

It looks like we have already handled this here - https://github.com/OE4T/meta-mender-community/pull/5, which exists in `meta-mender-community` dunfell branch - https://github.com/mendersoftware/meta-mender-community/blob/dunfell/meta-mender-tegra/recipes-bsp/u-boot/patches/0013-Reduce-env-size-on-p3450-0000-to-64KiB.patch (Only patches SPI/SD machines)

Looks like the nv_update_engine doesn't care about VER partitions being wiped - because mender updates work fine!

---

I think we need to apply above-mentioned patch on p3450_0002_defconfig as well. I added that just now. I tested and the mender update works fine!

Here's the layout after the fix:
```
$ hexdump mmcblk0boot1_fixed_mender -s 3911680 -n 64
03bb000 a8f7 5e3a 6102 746c 6f62 746f 6d63 3d64
03bb010 7572 206e 656d 646e 7265 615f 746c 6f62
03bb020 746f 6d63 3b64 7220 6e75 6220 6f6f 6374
03bb030 646d 6100 6372 3d68 7261 006d 6162 6475
03bb040
$ hexdump mmcblk0boot1_fixed_mender -s 3977216 -n 64
03cb000 7a4a 2852 6101 746c 6f62 746f 6d63 3d64
03cb010 7572 206e 656d 646e 7265 615f 746c 6f62
03cb020 746f 6d63 3b64 7220 6e75 6220 6f6f 6374
03cb030 646d 6100 6372 3d68 7261 006d 6162 6475
03cb040
$ hexdump mmcblk0boot1_fixed_mender -s 4063232 -n 64
03e0000 564e 0a33 2023 3352 2032 202c 4552 4956
03e0010 4953 4e4f 203a 2e34 0a33 4f42 5241 4944
03e0020 3d44 3433 3834 4220 414f 4452 4b53 3d55
03e0030 3030 3230 4620 4241 343d 3030 320a 3230
03e0040
$ hexdump mmcblk0boot1_fixed_mender -s 4128768 -n 64
03f0000 564e 0a33 2023 3352 2032 202c 4552 4956
03f0010 4953 4e4f 203a 2e34 0a33 4f42 5241 4944
03f0020 3d44 3433 3834 4220 414f 4452 4b53 3d55
03f0030 3030 3230 4620 4241 343d 3030 320a 3230
03f0040
```

ichergui commented on layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-bsp/u-boot/u-boot-tegra_%.bbappend:
```diff
@@ -0,0 +1,5 @@
+
+SRC_URI_append_mender-uboot = " file://0014-Reduce-env-size-on-p3450-0002-to-64-KiB.patch"
+
+TEGRA_MENDER_BOOTENV_SIZE_DEFAULT_tegra210 = "0x10000"
```
Please add new line here.

madisongh commented:
> Looks like the nv_update_engine doesn't care about VER partitions being wiped - because mender updates work fine!

nv_update_engine isn't used on the t210 machines (TX1 and Nano).  There's a Python script that NVIDIA provides for bootloader updates on those platforms.  The logic I have in tegra-boot-tools is more or less the same as what's implemented in that script.

The reason Mender updates work fine with TX1/Nano is because Mender updates don't do bootloader updates on the t120 platforms.  So you'd never notice this.

For an OTA reflash, we should be erasing the boot device if it needs to be reformatted - which it would if you've got the Uboot-env-overwriting-VER problem.  I added some improvements to tegra-boot-tools and the tegra-sysinstall scripts to better handle the needs-erasing case, but they'll also need an updated `nvflashxmlparse` script as well.  I've got the t-b-t and script updates in master currently, and the updated tegra-sysinstall in the master branch of my test distro.

atharvanan1 commented:
NOTE: this change doesn't work when you move from `dunfell-l4t-r32.4.3` to anything built from this branch on `jetson-nano-emmc` device. 

It breaks because VER partitions are overriden and `tegra-bootloader-update` complains about it.

My solution was to pack the VER partition contents into the new image and write them. However, @dwalkes mentioned UBENV location being changed - which might break things. @dwalkes do you want to jump in and give your input.

atharvanan1 commented:
> The reason Mender updates work fine with TX1/Nano is because Mender updates don't do bootloader updates on the t120 platforms. So you'd never notice this.

Thanks for your input @madisongh! Now I understand why we were using `--dry-run` with `tegra-bootloader-update`. If we aren't updating the bootloader -> should be continue to do that? Maybe add a verbose message about "Hey this is the error", but let the script go ahead without exiting?

madisongh commented:
> Thanks for your input @madisongh! Now I understand why we were using `--dry-run` with `tegra-bootloader-update`. If we aren't updating the bootloader -> should be continue to do that? Maybe add a verbose message about "Hey this is the error", but let the script go ahead without exiting?

Bootloader update support should probably be added, although it does introduce some risk, since a power failure during a bootloader update could brick the device.

atharvanan1 commented:
> Bootloader update support should probably be added, although it does introduce some risk, since a power failure during a bootloader update could brick the device.

Shouldn't we have the redundant partitions to take care of a power-failure during bootloader update? or you're specifically talking about "first" bootloader update?

Honestly, I'd let you make a decision on that, as you have more information on about lots of moving parts here. I'm happy to help and see through with the changes you want in - with your help of course.

dwalkes commented:
I think given JP 4.6 is the end of the road for Nano we ideally shouldn't invest the time in mender bootloader upgrade at this point.  I think nano users will have an OTA reflash option with this pull request (sort of) and maybe they will be lucky enough that the JP 4.4 bootloader works with the JP 4.6 image?  

The thing I'm still not sure about his how to handle the dunfell-l4t-r32.4.3 branch.  If we introduce the changes here which move uboot environment for OTA support I think this means we will potentially break mender updates for any nano users who are already on this branch since `tegra-bootloader-update` won't like their corrupted VER partitions.  Moving uboot environment silently when you mender upgrade to the latest branch seems scary to me too and I'm worried this will be difficult to test and prove is working.

madisongh commented:
> I think given JP 4.6 is the end of the road for Nano we ideally shouldn't invest the time in mender bootloader upgrade at this point. I think nano users will have an OTA reflash option with this pull request (sort of) and maybe they will be lucky enough that the JP 4.4 bootloader works with the JP 4.6 image?

If we're talking about the SDcard-based dev kit, then maybe it's not a big deal - nobody should be using that model in a production environment.  The main reason for integrating BUP handling into Mender updates for the t210 machines is to make sure the device tree matches the kernel. It's less likely (though still possible) that some of the other firmware loaded at boot time might change in a way that could cause weird problems if there's a mismatch. (But then there's the rollback problem, where if something does change in an incompatible way, but the new OS doesn't finish booting, failing back over to the old rootfs wouldn't necessarily work, either.)

There was a big change with the bootloader and flash layout for the SDcard-based Nanos between 4.4/32.4.3 and 4.5/32.5.0, so a reflash is more or less required past 4.4/32.4.3 on that platform.  That change doesn't affect the eMMC-based Nanos, though, and since those are more likely to be used in real products, having a full update solution that works for them would be a plus.

> Moving uboot environment silently when you mender upgrade to the latest branch seems scary to me too and I'm worried this will be difficult to test and prove is working.

Saving the values of the Mender U-Boot variables before applying the update, then restoring them after the update is in place, should work OK, but yes, difficult to verify for the general case.  It should be part of the OTA re-flash process, though.

atharvanan1 commented:
> It's less likely (though still possible) that some of the other firmware loaded at boot time might change in a way that could cause weird problems if there's a mismatch. (But then there's the rollback problem, where if something does change in an incompatible way, but the new OS doesn't finish booting, failing back over to the old rootfs wouldn't necessarily work, either.)

Hey @madisongh, @dwalkes and I talked about this and we think that it would be more work to take on this for the scope of this pull-requests. 

Here's a solution for the `jetson-nano-emmc`:
* Let's not make the partition change so that we continue to override the VER partitions. Give a verbose message and link people to this thread for additional information. (`jetson-nano-emmc` will continue to work as it has been)
    * In tegra-bootloader-update, we can add --ignore-ver-check argument and add that [here](https://github.com/BoulderAI/tegra-demo-distro/blob/e8523902690b2695def2a634b549d3aeeb51f7a9/layers/meta-tegra-support/dynamic-layers/meta-mender-tegra/recipes-mender/tegra-state-scripts/files/redundant-boot-install-script-uboot#L63) after it fails first because of corrupted VER partitions
* A warning for anyone coming from JP4.4 and wanting to upgrade to JP4.6 that bup updates are not supported yet and might need additional work and testing to make sure it works.
* A warning for all machines that we will be using `tegra-boot-tools` onwards for supporting reflashing OTA.

atharvanan1 commented:
I have a change in place to remove the partition layout change, and updates to the script to add verbose messaging in a local branch. I'll run some tests and update you on that!

atharvanan1 commented:
Log showing the warning for `jetson-nano-emmc` device:

```root@jetson-nano-emmc:~# mender -install http://10.1.10.120:8000/demo-image-base-jetson-nano-emmc.mender
INFO[0000] Loaded configuration file: /var/lib/mender/mender.conf 
INFO[0000] Loaded configuration file: /etc/mender/mender.conf 
INFO[0000] Mender running on partition: /dev/mmcblk0p1  
INFO[0000] Performing remote update from: [http://10.1.10.120:8000/demo-image-base-jetson-nano-emmc.mender]. 
Installing Artifact of size 196520960...
INFO[0000] No public key was provided for authenticating the artifact 
INFO[0000] Update Module path "/usr/share/mender/modules/v3" could not be opened (open /usr/share/mender/modules/v3: no such file or directory). Update modules will not be available 
INFO[0000] Opening device "/dev/mmcblk0p18" for writing 
INFO[0000] Native sector size of block device /dev/mmcblk0p18 is 512 bytes. Mender will write in chunks of 1048576 bytes 
.............................................................. - 100 %
INFO[0036] All bytes were successfully written to the new partition 
INFO[0036] The optimized block-device writer wrote a total of 4743 frames, where 0 frames did need to be rewritten (i.e., skipped) 
INFO[0037] Wrote 4972347392/4972347392 bytes to the inactive partition 
INFO[0037] Enabling partition with new image installed to be a boot candidate: 18 
INFO[0038] Executing script: ArtifactInstall_Leave_80_bl-update 
INFO[0038] Collected output (stderr) while running script /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update
WARN: VER partitions are corrupted
WARN: Please refer to https://github.com/OE4T/tegra-demo-distro/pull/113 for additional context
WARN: Attempting update anyway

---------- end of script output 
Use -commit to update, or -rollback to roll back the update.
At least one payload requested a reboot of the device it updated.
```

Will run some mender torture on this device. Pushing cleaned commits.

atharvanan1 commented:
> For an OTA reflash, we should be erasing the boot device if it needs to be reformatted - which it would if you've got the Uboot-env-overwriting-VER problem. I added some improvements to tegra-boot-tools and the tegra-sysinstall scripts to better handle the needs-erasing case, but they'll also need an updated `nvflashxmlparse` script as well. I've got the t-b-t and script updates in master currently, and the updated tegra-sysinstall in the master branch of my test distro.

I have something in works to add these changes into `dunfell-l4t-r32.4.3` branch from `meta-tegra` [here](https://github.com/BoulderAI/meta-tegra/tree/dunfell-l4t-r32.4.3%2Btbt-nvflashxml-script-changes)

---

Tested mender updates on `jetson-nano-emmc` it works fine. Couldn't do full mender-torture, as it didn't work for some reason. What did I check?

* Mender update changes the rootfsA to rootfsB.
* Rebooting without `mender -commit` boots back to rootfsA. It stays like that for at least 3 reboots (that's what I tested)
* Rebooting with `mender -commit` boots into rootfsB. It stays like that for at least 3 reboots

dwalkes commented:
> Couldn't do full mender-torture, as it didn't work for some reason. 

I don't think we support mender-torture on u-boot environments yet, see https://github.com/mendersoftware/meta-mender-community/tree/dunfell/meta-mender-tegra/scripts/test#tegra-mender-torture-tests

atharvanan1 commented:
Tested this today with latest tegra-boot-tools changes. Updated the README with a notice for change in boot-tools.

madisongh commented:
@atharvanan1 LGTM now.... is there anything else that needs to be checked before I merge?

atharvanan1 commented:
Let's sit on it for a day, I'll think if I'm missing anything.

atharvanan1 commented:
Here's an overview of what's changed and tested:

1. `tegra-redundant-tools` are removed and replaced with `tegra-boot-tools-updater`. These are necessary for OTA reflashing update, given that `full_init_bl_payload` is present which is added by [this commit](https://github.com/OE4T/meta-tegra/commit/bdad3ebf89224e5c3d048835a24219c134ff1d60). Updated the README for that.
2. Mender scripts have been changed for all devices to use `tegra-boot-tools-updater`. For this, mender torture has been tested on `jetson-tx2` and `jetson-xavier-nx-devkit-emmc` devices. For `jetson-nano-emmc` device, we give an verbose message if VER partitions are corrupted - but continue with the update anyway. Tested on `jetson-nano-emmc` as [per this comment](https://github.com/OE4T/tegra-demo-distro/pull/113#issuecomment-864329118).
3. Reflashing process has been tested on Jetson TX2 as per https://github.com/OE4T/meta-tegra/wiki/Over-the-air-reflashing-process.

Let me know if I should test any other devices (SD Card based devices), however as I understand they should work as their emmc counterparts work.

madisongh commented:
Thanks for all the work on this, @atharvanan1 !

atharvanan1 commented:
Thanks for helping me on this! @madisongh
